alternator/streams.cc

```diff
@@ -479,50 +476,135 @@ future<executor::request_return_type> executor::describe_stream(client_state& cl
-        auto i = topologies.begin();
+
+        // filter out cdc generations older than the table or now() - dynamo_streams_max_window (24h)
+        auto low_ts = std::max(as_timepoint(schema->id()), db_clock::now() - dynamo_streams_max_window);
+
+        auto i = topologies.lower_bound(low_ts);
+        // need first gen _intersecting_ the timestamp.
+        if (i != topologies.begin()) {
+            --i;
+        }
+
         auto e = topologies.end();
         auto prev = e;
-        auto threshold = db_clock::now() - confidence_interval(db);
         auto shards = rjson::empty_array();
 
         std::optional<shard_id> last;
 
+        // i is now at the youngest generation we include. make a mark of it.
+        auto first = i;
+
+        // if we're a paged query, skip to the generation where we left of.
+        if (shard_start) {
+            i = std::find_if(i, e, [&](auto& p) {
+                return p.first == shard_start->start_time;
+            });
+        }
+
+        // comparator to sort cdc::stream_id:s by index and token
+        auto cmp = [](const cdc::stream_id& id1, const cdc::stream_id& id2) {
+            auto i1 = id1.index();
+            auto i2 = id2.index();
+            return i1 < i2 || (i1 == i2 && id1.token() < id2.token());
+        };
+
+        // need a prev even if we are skipping stuff
+        if (i != first) {
+            auto j = i;
+            --j;
+            std::sort(j->second.streams().begin(), j->second.streams().end(), cmp);
+            prev = j;
+        }
+
         for (; limit > 0 && i != e; prev = i, ++i) {
             auto& [ts, sv] = *i;
 
             last = std::nullopt;
 
-            for (auto& id : sv.streams()) {
-                if (shard_start && shard_start->id != id) {
-                    continue;
+            // TODO: get rid of this once cdc generations
+            // management starts correctly updating
+            // cdc expiration columns in topology/desc
+            // or the "expired" part of topology
+            // is retired.
+            auto get_expiration = [&](auto iter) {
+                auto exp = iter->second.expired();
+                if (!exp && ++iter != e) {
+                    exp = iter->first;
                 }
-                if (shard_start && shard_start->id == id) {
-                    shard_start = std::nullopt;
-                    continue;
+                return exp.value_or(db_clock::time_point{});
+            };
+
+            auto lo = sv.streams().begin();
+            auto end = sv.streams().end();
+
+            // operating on a copy. ok to sort. this order is also preserved 
+            // for the "next" check below.
+            std::sort(lo, end, cmp);
+
+            assert(std::is_sorted(std::find_if(lo, end, [](auto& id) { return id.index() == 1; }), end, [](auto& id1, auto& id2) {
```

Comment:
Probably best to avoid generic lambdas.

Reply:
Again, based on the size of the lambda, me thinks it is warranted.